From c901a0d9830a509e544372747b5bd5811b1c15f2 Mon Sep 17 00:00:00 2001
From: Sunny Goyal <sunnygoyal@google.com>
Date: Thu, 13 Feb 2025 09:49:26 -0800
Subject: [PATCH] Handle exceptions from querying appinfo in
 RemoteViews#addAppWidget.

Host process may not have access to the ApplicationInfo directly in some cases

Bug: 395168279
Change-Id: Ic26d63acea5f227b56d44bc2e417f7b189f0d2f2
Test: Manual
Flag: EXEMPT bugfix
(cherry picked from commit 37bf5823504f2a256f128123393cd149721b87fc)
---
 core/java/android/widget/RemoteViews.java | 16 ++++++++++------
 1 file changed, 10 insertions(+), 6 deletions(-)

diff --git a/core/java/android/widget/RemoteViews.java b/core/java/android/widget/RemoteViews.java
index da7c21c826b6..4948ff52bafb 100644
--- a/core/java/android/widget/RemoteViews.java
+++ b/core/java/android/widget/RemoteViews.java
@@ -6068,12 +6068,16 @@ public class RemoteViews implements Parcelable, Filter {
                 return context;
             }
             try {
-                // Use PackageManager as the source of truth for application information, rather
-                // than the parceled ApplicationInfo provided by the app.
-                ApplicationInfo sanitizedApplication =
-                        context.getPackageManager().getApplicationInfoAsUser(
-                                mApplication.packageName, 0,
-                                UserHandle.getUserId(mApplication.uid));
+                ApplicationInfo sanitizedApplication = mApplication;
+                try {
+                    // Use PackageManager as the source of truth for application information, rather
+                    // than the parceled ApplicationInfo provided by the app.
+                    sanitizedApplication = context.getPackageManager().getApplicationInfoAsUser(
+                        mApplication.packageName, 0, UserHandle.getUserId(mApplication.uid));
+                } catch(SecurityException se) {
+                    Log.d(LOG_TAG, "Unable to fetch appInfo for " + mApplication.packageName);
+                }
+
                 Context applicationContext = context.createApplicationContext(
                         sanitizedApplication,
                         Context.CONTEXT_RESTRICTED);
-- 
2.50.0.rc0.604.gd4ff7b7c86-goog

